src/transformations.rs

```diff
@@ -368,14 +397,32 @@ impl Closurizable for RichTerm {
     /// Generate a fresh variable, bind it to the corresponding closure `(t,with_env)` in `env`,
     /// and return this variable as a fresh term.
     fn closurize(self, env: &mut Environment, with_env: Environment) -> RichTerm {
+        // If the term is already a generated variable (that is, introduced by the share normal
+        // form transformation), we don't have to create a useless intermediate closure. We just
+        // transfer the original thunk to the new environment. This is not only an optimization:
+        // this is relied upon by recursive record merging when computing the fixpoint. Change
+        // carefully.
+        //
+        // We could do it for non-generated ident as well, but be we would be renaming
+        // user-supplied variables by giberrish generated names. It may hamper error reporting, so
+        // for the time being, we restrict ourselves to generated identifier.
+        let reuse_thunk = match self.as_ref() {
+            Term::Var(id) if id.is_generated() => with_env.get(&id),
+            _ => None,
+        };
+
         let var = fresh_var();
         let pos = self.pos;
-        let closure = Closure {
-            body: self,
-            env: with_env,
-        };
-        env.insert(var.clone(), Thunk::new(closure, IdentKind::Record()));
 
+        let thunk = reuse_thunk.unwrap_or_else(|| {
+            let closure = Closure {
+                body: self,
+                env: with_env,
+            };
+            Thunk::new(closure, IdentKind::Record())
+        });
```

Comment:
I would simply move all of this in the `match` above, and avoid indirecting through `Option`.

Reply:
Initially it's because in case the thunk is not found in the environment, I wanted to fallback on the same code (418 - 422). If we were to put it in the match, we would have to duplicate it (`with_env.get` returns an `Option`). But now that I think of it, not finding a generated identifier in the environment is something we want to hard fail on, and the fallback does violate the invariant relied upon by the evaluation of recursive records mentioned in the comment above. So let's rather panic! or bubble an `Err`